service/controller/resource/deployment/update.go

```diff
@@ -73,49 +130,69 @@ func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desir
+		desiredDeployment, err := getDeploymentByName(desiredDeployments, currentDeployment.Name)
+		if IsNotFound(err) {
+			// NOTE that this case indicates we should remove the current deployment
+			// eventually.
+			r.logger.LogCtx(ctx, "level", "warning", "message", fmt.Sprintf("not updating deployment '%s': no desired deployment found", currentDeployment.GetName()))
+			continue
+		} else if err != nil {
+			return nil, microerror.Mask(err)
 		}
 
-		// We select one deployment to be updated per reconciliation loop. Therefore
-		// we have to check its state on the version bundle level to see if a
-		// deployment is already up to date. We also check if there are any other
-		// changes on the pod specs. In case there are none, we check the next one.
-		// The first one not being up to date will be chosen to be updated next and
-		// the loop will be broken immediatelly.
-		for _, currentDeployment := range currentDeployments {
-			desiredDeployment, err := getDeploymentByName(desiredDeployments, currentDeployment.Name)
-			if IsNotFound(err) {
-				// NOTE that this case indicates we should remove the current deployment
-				// eventually.
-				r.logger.LogCtx(ctx, "level", "warning", "message", fmt.Sprintf("not updating deployment '%s': no desired deployment found", currentDeployment.GetName()))
-				continue
-			} else if err != nil {
-				return nil, microerror.Mask(err)
-			}
+		if !isDeploymentModified(desiredDeployment, currentDeployment) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("not updating deployment '%s': no changes found", currentDeployment.GetName()))
+			continue
+		}
 
-			if !isDeploymentModified(desiredDeployment, currentDeployment) {
-				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("not updating deployment '%s': no changes found", currentDeployment.GetName()))
-				continue
+		// If worker deployment, check that master does not have any prohibited states before updating the worker
+		if tcK8sClient != nil {
+			if desiredDeployment.ObjectMeta.Labels[key.LabelApp] == key.WorkerID {
+				// List all master nodes in the tenant
+				tcNodes, err := tcK8sClient.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: "role=master"})
+				if err != nil {
+					r.logger.LogCtx(ctx, "level", "warning", "message", "unable to list tenant cluster master nodes")
+					return nil, microerror.Mask(err)
+				}
+				for _, n := range tcNodes.Items {
+					if key.NodeHasNoScheduleOrNoExecute(n) {
+						// Node has NoSchedule or NoExecute taint
+						msg := fmt.Sprintf("not updating deployment '%s': one or more tenant cluster master nodes are unschedulable", currentDeployment.GetName())
+						r.logger.LogCtx(ctx, "level", "warning", "message", msg)
+						continue DeploymentsLoop
+					}
+				}
 			}
+		} else {
+			r.logger.LogCtx(ctx, "level", "warning", "message", "unable to check tenant cluster master status. No tenant cluster client configured")
+			continue DeploymentsLoop
```

Comment:
I think this could be a plain `continue` since it's not in a nested `for`.

Reply:
It could be, but I wanted it to be the same as the other continue to be explicit that they both continue to the outer loop. Open to changing it if you have a strong feeling about it

Reply:
Actually I removed this altogether as the nil check isn't needed anymore